modules/service-livepatch.sh

```diff
@@ -3,25 +3,92 @@
 LIVEPATCH_SERVICE_TITLE="Canonical Livepatch"
 LIVEPATCH_SUPPORTED_SERIES="trusty xenial bionic"
 LIVEPATCH_SUPPORTED_ARCHS="x86_64"
+LIVEPATCH_FALLBACK_KERNEL="linux-image-generic"
+
+_livepatch_install_supported_kernel() {
+    if apt_is_package_installed "${LIVEPATCH_FALLBACK_KERNEL}"; then
+        return 0
+    fi
+    echo 'A Livepatch compatible kernel will be installed.'
+    echo -n 'Running apt-get update... '
+    check_result apt_get update
+    echo -n "Installing ${LIVEPATCH_FALLBACK_KERNEL}... "
+    check_result apt_get install "${LIVEPATCH_FALLBACK_KERNEL}"
+}
+
+_livepatch_try_enable() {
+    local token="$1"
+    local allow_kernel_change="$2"
+    local output=""
+    local result=0
+    local disabled_reason=""
+
+    output=$(canonical-livepatch enable "$token" 2>&1) || result=$?
+    if [ "${result}" -eq "0" ]; then
+        echo "${output}"
+        return 0
+    fi
+    # ok, we failed, why?
+    disabled_reason=$(livepatch_disabled_reason "${output}")
+    if echo "${disabled_reason}" | grep -q "unsupported kernel"; then
+        echo -n "Your running kernel ${KERNEL_VERSION} is not supported by "
+        echo 'Livepatch.'
+        if [ "${allow_kernel_change}" = "yes" ]; then
+            _livepatch_install_supported_kernel
+            echo 'A new kernel was installed to support Livepatch.'
+            echo 'Please reboot into it and then run the enable command again:'
+            echo
+            echo "sudo $SCRIPTNAME enable-livepatch $token"
+        else
+            echo
+            echo -n 'If you want to automatically install a Livepatch supported '
+            echo 'kernel, please run:'
+            echo -n "sudo $SCRIPTNAME enable-livepatch $token "
+            echo '--allow-kernel-change'
+        fi
+        error_exit livepatch_unsupported_kernel
+    else
+        error_msg "${output}"
+        exit ${result}
+    fi
+}
+
+_old_kernel_message() {
+    local token="$1"
+
+    echo
+    echo "Your currently running kernel ($KERNEL_VERSION) is too old to"
+    echo "support snaps. Version 4.4.0 or higher is needed."
```

Comment:
The user asked for livepatch, and doesn't know what snaps are or what they have to do with livepatch. What's this message about snaps? :)

Reply:
I could clarify that Livepatch is delivered as a snap, and snaps require a newer kernel. Note this is not a message introduced in this branch, I just moved it to its own function to make the code easier to read.

I grant that there might be some confusion between this kernel update for snaps in general, and the other kernel *change* for livepatch support. Unfortunately I cannot determine if a kernel is suitable for livepatching before I install the canonical-livepatch snap. That snap is what tells me if the running kernel can be livepatched or not.